tests/hooks/test_subprocess.py

```diff
@@ -0,0 +1,110 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import unittest
+from pathlib import Path
+from subprocess import PIPE, STDOUT
+from tempfile import TemporaryDirectory
+from unittest import mock
+
+import pytest
+from parameterized import parameterized
+
+from airflow.exceptions import AirflowException
+from airflow.hooks.subprocess import SubprocessHook
+
+OS_ENV_KEY = 'SUBPROCESS_ENV_TEST'
+OS_ENV_VAL = 'this-is-from-os-environ'
+
+
+class TestSubprocessHook(unittest.TestCase):
+    @parameterized.expand(
+        [
+            ('with env', {'ABC': '123', 'AAA': '456'}, {'ABC': '123', 'AAA': '456', OS_ENV_KEY: ''}),
+            ('empty env', {}, {OS_ENV_KEY: ''}),
+            ('no env', None, {OS_ENV_KEY: OS_ENV_VAL}),
+        ]
+    )
+    def test_env(self, name, env, expected):
+        """
+        Test that env variables are exported correctly to the command environment.
+        When ``env`` is ``None``, ``os.environ`` should be passed to ``Popen``.
+        Otherwise, the variables in ``env`` should be available, and ``os.environ`` should not.
+        """
+        print(f"test_name: {name}")
+        hook = SubprocessHook()
+
+        def build_cmd(keys, filename):
+            """
+            Produce bash command to echo env vars into filename.
+            Will always echo the special test var named ``OS_ENV_KEY`` into the file to test whether
+            ``os.environ`` is passed or not.
+            """
+            return '\n'.join([f"echo {k}=${k}>> {filename}" for k in [*keys, OS_ENV_KEY]])
+
+        with TemporaryDirectory() as tmp_dir, mock.patch.dict('os.environ', {OS_ENV_KEY: OS_ENV_VAL}):
+            tmp_file = Path(tmp_dir, 'test.txt')
+            command = build_cmd(env and env.keys() or [], tmp_file.as_posix())
+            hook.run_command(command=['bash', '-c', command], env=env)
+            actual = dict([x.split('=') for x in tmp_file.read_text().splitlines()])
+            assert actual == expected
+
+    @parameterized.expand(
+        [
+            ('test-val', 'test-val'),
+            ('test-val\ntest-val\n', ''),
```

Comment:
This case looks odd - I would have expected "last line" here to be `test-val`, not empty.

Reply:
Existing behaviors to return the last line in standard out, not the last non-empty line.

I thought about maybe tweaking this behavior in the hook for example returning all output and letting operators decide what to do with it, or somehow returning a generator or storing the full output on a instance attribute. But I figure these things could be done later. Lmk if you think something like that is worth pursuing now

Reply:
Existing behaviors to return the last line in standard out, not the last non-empty line.

I thought about maybe tweaking this behavior in the hook for example returning all output and letting operators decide what to do with it, or somehow returning a generator or storing the full output on a instance attribute. But I figure these things could be done later. Lmk if you think something like that is worth pursuing now

Reply:
Oh right, if this is existing behaviour 👍🏻